Adding guidelines for writing platform agnostic code (#18188)
open-telemetry/opentelemetry-collector-contrib: CONTRIBUTING.md

Alternately, copy `./.chloggen/TEMPLATE.yaml`, or just create your file from scratch.

> ## Platform Agnostic Code
> 
> In order to ensure compatibility with different operating systems, code should be platform-agnostic. Below are some guidelines to follow when writing platform-agnostic code:

Reviewer: ```suggestion
## Portable Code

In order to ensure compatibility with different operating systems, code should be portable. Below are some guidelines to follow when writing portable code:
```

I don't think the goal should be to be platform agnostic, but instead to be portable.  Some of the things we're doing are highly platform dependent.

> In order to ensure compatibility with different operating systems, it is advised to write a code that is platform-agnostic. Below are some guidelines to follow when writing platform-agnostic code:

Reviewer 2: ```suggestion
In order to ensure compatibility with different operating systems code should be platform-agnostic. Below are some guidelines to follow when writing platform-agnostic code:
```

> * Avoid using platform-specific libraries, features etc. Please opt for cross-platform solutions. 

Reviewer: This doesn't seem to say anything of substance.  What is the actionable recommendation here?  What is a `cross-platform` solution?  Is a portable multi-platform solution that uses different implementations per platform acceptable?  That would not be platform agnostic.

> * Avoid hard-coding platform specific values. Use environment variables or configuration files for storing platform specific values.

Reviewer: What sort of platform-specific values are you expecting here?  Examples of what to do and what not to do would be helpful.

Author: Added an example, let me know if that is helpful.

> * Avoid hard-coding platform specific values. Use environment variables or configuration files for storing platform specific values.

Reviewer: ```suggestion
* Avoid hard-coding platform-specific values. Use environment variables or configuration files for storing platform-specific values.
```

* Be mindful of 

>   - Standard file systems and file paths such as forward slashes (/) instead of backward slashes (\) in Windows. 

Reviewer: Can we provide specific guidance here such as to use the [`path/filepath` package](https://pkg.go.dev/path/filepath) when working with file paths?

  - Consistent line ending formats such as Unix (LF) or Windows (CRLF).

> * Test your implementation thoroughly on different platforms and fix any issues. 

Reviewer 2: I agree with this statement, but the Contrib CI is currently not enforcing Windows tests with every build :/ Doesn't feel right to recommend it when we aren't doing it ourselves.  @open-telemetry/collector-contrib-approvers what do you think?

Reviewer: I think that when working on platform-dependent code (which I expect will continue to exist, even as part of a portable implementation) it is important to test on the relevant platforms.  That said, not all developers will have all platforms available to them and the CI matrix is important to provide assurance that we're consistently testing on supported platforms.

Author: @Aneurysm9 Should we keep the testing line or remove it for now?

Reviewer: v0v

Really not sure what the right answer here is.  If multi-platform testing is needed it is needed, but for many PRs it won't be and the author can rely on the CI system to do it for them.

Author: How about asking the best effort testing ? Like below

`Test your implementation thoroughly on different platforms if possible and fix any issues.`

Alternately, copy `./.chloggen/TEMPLATE.yaml`, or just create your file from scratch.

## Portable Code

In order to ensure compatibility with different operating systems, code should be portable. Below are some guidelines to follow when writing portable code:

* Avoid using platform-specific libraries, features etc. Please opt for portable multi-platform solutions. 

* Avoid hard-coding platform-specific values. Use environment variables or configuration files for storing platform-specific values.

    For example, avoid using hard-coded file path
    ```

>     filePath := "C:/Users/Bob/Documents/sampleData.csv"

Reviewer: ```suggestion
    filePath := "C:\Users\Bob\Documents\sampleData.csv"
```

Windows path separator is `\`, unless you were trying to illustrate getting it wrong by assuming all platforms use `/`.

Author: Fixed, wasn't intentionally making it wrong, just a typo.

    ```

    Instead environment variable or configuration file can be used.
    ```
    filePath := os.Getenv("DATA_FILE_PATH")
    ```
    or
    ```
    filePath := Configuration.Get("data_file_path")
    ```

* Be mindful of 

>   - Standard file systems and file paths such as forward slashes (/) instead of backward slashes (\\) in Windows. Refer to [path/filepath package](https://pkg.go.dev/path/filepath) when working with filepaths. 

Reviewer: ```suggestion
  - Standard file systems and file paths such as backslashes (\\) instead of slashes (/) in Windows. Use the [`path/filepath` package](https://pkg.go.dev/path/filepath) when working with filepaths. 
```
